runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,267 @@
+    file: Arc<Mutex<(File, IoUring, u64)>>,
+}
+
+impl Blob {
+    pub fn new(partition: String, name: &[u8], file: File, len: u64) -> Self {
+        let ring = IoUring::new(32).unwrap();
+        Self {
+            partition,
+            name: name.into(),
+            file: Arc::new(Mutex::new((file, ring, len))),
+        }
+    }
+}
+
+impl crate::Blob for Blob {
+    async fn len(&self) -> Result<u64, Error> {
+        let inner = self.file.lock().map_err(|_| Error::LockGrabFailed)?;
+        let (_, _, len) = &*inner;
+        Ok(*len)
+    }
+
+    async fn read_at(&self, buf: &mut [u8], offset: u64) -> Result<(), Error> {
+        let mut inner = self.file.lock().map_err(|_| Error::LockGrabFailed)?;
+        let (file, ring, len) = &mut *inner;
+
+        if offset + buf.len() as u64 > *len {
+            return Err(Error::BlobInsufficientLength);
+        }
+
+        let fd = types::Fd(file.as_raw_fd());
+        let mut total_read = 0;
+
+        while total_read < buf.len() {
+            let remaining = &mut buf[total_read..];
+
+            // Prepare the read operation
+            let read_e = opcode::Read::new(fd, remaining.as_mut_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation to the ring
+            unsafe {
+                ring.submission()
+                    .push(&read_e)
+                    .map_err(|_| Error::ReadFailed)?;
+            }
+
+            // Wait for the operation to complete
+            ring.submit_and_wait(1).map_err(|_| Error::ReadFailed)?;
+
+            // Process the completion event
+            let cqe = ring.completion().next().ok_or(Error::ReadFailed)?;
+
+            // If the return value is non-positive, it indicates an error.
+            let bytes_read: usize = cqe.result().try_into().map_err(|_| Error::ReadFailed)?;
+            if bytes_read == 0 {
+                // Got EOF before filling buffer.
+                return Err(Error::BlobInsufficientLength);
+            }
+
+            total_read += bytes_read;
+        }
+
+        Ok(())
+    }
+
+    async fn write_at(&self, buf: &[u8], offset: u64) -> Result<(), Error> {
+        let mut inner = self.file.lock().map_err(|_| Error::LockGrabFailed)?;
+        let (file, ring, len) = &mut *inner;
+
+        let fd = types::Fd(file.as_raw_fd());
+        let mut total_written = 0;
+
+        while total_written < buf.len() {
+            let remaining = &buf[total_written..];
+
+            // Prepare the write operation
+            let write_op = opcode::Write::new(fd, remaining.as_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation to the ring
+            unsafe {
+                ring.submission()
+                    .push(&write_op)
+                    .map_err(|_| Error::WriteFailed)?; // TODO danlaine: consider changing error values.
```

Comment:
Per this TODO, and related to my comment above: do we want our errors to carry more context than this? Or is this OK for now?